### DISCUSSION TITLE
Better matching of Requested PGN for PGN 126208 to Field Definitions

### DISCUSSION THREAD

**Discussion by cas206** [canboat/canboat#440] [Ideas]
PGN 126208 can send or request parameters for the PGN of another destination.  The following example output is requesting fields 1, 3, 4, and 5 from PGN 126720.  The default output from analyzer is:

2024-02-17T05:15:49.768Z 3 204 172 126208 NMEA - Request group function:  Function Code = Request; PGN = 126720; Transmission interval = Unknown; Transmission interval offset = Unknown; Number of Parameters = 4; Parameter 1 = 1; Value 1 = Raymarine; Parameter 2 = 3; Value 2 = Marine Industry; Parameter 3 = 4; Value 3 = 92 =240; Parameter 4 = 5; Value 4 = 15 =129

The code uses the PGN definition list to determine the data format details of the requested fields in the requested PGN.  The reported values looks incorrect, but the code is working as designed.  The snippets in question are: 
“Parameter 3 = 4; Value 3 = 92 =240;" and 
"Parameter 4 = 5; Value 4 = 15 =129”

The function fieldPrintVariable() at line 1210 of analyzer.c calls getField() at line 1220.  getField() currently only returns the first match from the PGN list.  However, the PGN list contains multiple versions of the requested PGN 126720 depending on the values in fields 1, 3, 4 and 5.  The first PGN match happens to contain the match type field values "Raymarine", "Marine Industry", “=240” and “=129”. The values reported in this example are "Raymarine", "Marine Industry", decimal 92 and decimal 15.  So the output shows “Value 3 = 92 =240” and “Value 4 = 15 =129”.  Note that, currently there is no definition of this variant of PGN 126720 that contains fields 1, 3, 4, and 5, of “Raymarine”, “Marine”, 92, and 15.  

This discussion proposes that a change be implemented to analyzer.c so that:
- It will search all match type fields when getting the definition of the requested PGN, instead of just getting the first matching PGN number.  Assuming that the above example had a match in the PGN List, it would show output such as “Value 3 = 92 = 92”.
- If a match of field values isn’t possible, that some alternative output be used, such as:
“Value 3 = 92 NoMatch”, “Value 3 = 92”,  “Value 3 = 92 (NA)”, or something like this.

cas

**Comment by keesverruijt** [comment]
Good suggestion.
